lib/glimesh_web/controllers/user_session_controller.ex

```diff
@@ -12,7 +13,17 @@ defmodule GlimeshWeb.UserSessionController do
     %{"email" => email, "password" => password} = user_params
 
     if user = Accounts.get_user_by_email_and_password(email, password) do
-      UserAuth.log_in_user(conn, user, user_params)
+      if user.tfa_token do
+        %{"tfa" => tfa} = user_params
+
+        if Tfa.validate_pin(tfa, user.tfa_token) do
+          UserAuth.log_in_user(conn, user, user_params)
+        else
+          render(conn, "new.html", error_message: "Invalid 2FA code")
+        end
+      else
+        UserAuth.log_in_user(conn, user, user_params)
+      end
```

Comment:
This is a lot of nesting, how can we flatten it a bit?

Reply:
What I can do to change this is change where I figure if you have a token and pass or fail so instead of
```elixir
if user.tfa_token do
  %{"tfa" => tfa} = user_params

  if Tfa.validate_pin(tfa, user.tfa_token) do
    UserAuth.log_in_user(conn, user, user_params)
  else
    render(conn, "new.html", error_message: "Invalid 2FA code")
  end
else
  UserAuth.log_in_user(conn, user, user_params)
end
```
it could be 
```elixir
if Tfa.validate_pin(tfa, user.tfa_token) do
  UserAuth.log_in_user(conn, user, user_params)
else
  render(conn, "new.html", error_message: "Invalid 2FA code")
end```
where the check is inside the validate_pin